envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/TlsBasedAuthenticationTest.kt

```diff
@@ -0,0 +1,226 @@
+
+    companion object {
+
+        private val properties = mapOf(
+            "envoy-control.envoy.snapshot.incoming-permissions.enabled" to true,
+            "envoy-control.envoy.snapshot.routes.status.create-virtual-cluster" to true,
+            "envoy-control.envoy.snapshot.routes.status.path-prefix" to "/status/",
+            "envoy-control.envoy.snapshot.routes.status.enabled" to true
+        )
+
+        private lateinit var echo1Envoy: EnvoyContainer
+        private lateinit var echo2Envoy: EnvoyContainer
+        private lateinit var echo3Envoy: EnvoyContainer
+
+        @JvmStatic
+        @BeforeAll
+        fun setupTest() {
+            setup(appFactoryForEc1 = { consulPort ->
+                EnvoyControlRunnerTestApp(properties = properties, consulPort = consulPort)
+            }, envoyConfig = Echo1EnvoyAuthConfig, secondEnvoyConfig = Echo2EnvoyAuthConfig, envoys = 2)
+            echo1Envoy = envoyContainer1
+            echo2Envoy = envoyContainer2
+        }
+
+        private fun registerEcho2WithEnvoyOnIngress() {
+            registerService(
+                    id = "echo2",
+                    name = "echo2",
+                    container = echo2Envoy,
+                    port = EnvoyContainer.INGRESS_LISTENER_CONTAINER_PORT,
+                    tags = listOf("mtls:enabled")
+            )
+        }
+
+        private fun registerEcho2EnvoyAsEcho3() {
+            registerService(
+                    id = "echo3",
+                    name = "echo3",
+                    container = echo2Envoy,
+                    port = EnvoyContainer.INGRESS_LISTENER_CONTAINER_PORT,
+                    tags = listOf("mtls:enabled")
+            )
+        }
+    }
+
+    @BeforeEach
+    fun setup() {
+        registerService(name = "echo", tags = listOf("mtls:enabled"))
+        registerEcho2WithEnvoyOnIngress()
+    }
+
+    @Test
+    fun `should encrypt traffic between selected services`() {
+        untilAsserted {
+            // when
+            val validResponse = callEcho2FromEcho1()
+
+            // then
+            val sslHandshakes = echo1Envoy.admin().statValue("cluster.echo2.ssl.handshake")?.toInt()
+            assertThat(sslHandshakes).isGreaterThan(0)
+
+            val sslConnections = echo2Envoy.admin().statValue("http.ingress_https.downstream_cx_ssl_total")?.toInt()
+            assertThat(sslConnections).isGreaterThan(0)
+
+            assertThat(validResponse).isOk().isFrom(echoContainer2)
+        }
+    }
+
+    @Test
+    fun `should not allow traffic that fails client SAN validation`() {
+        echo3Envoy = createEnvoyContainerWithEcho3San()
+        echo3Envoy.start()
+
+        untilAsserted {
+            // when
+            // echo2 doesn't allow requests from echo3
+            val invalidResponse = callEcho2FromEcho3()
+
+            // then
+            val sanValidationFailure = echo2Envoy.admin().statValue("http.ingress_https.rbac.denied")?.toInt()
```

Comment:
Should we check `rbac` metric? 
Maybe we should check ssl metrics like in test below.

Reply:
SSL connection was established and everything was fine, the only thing that did not work here was that in RBAC echo3 was not allowed. There is no ssl metric that would show that. Am I right @MarcinFalkowski ?

Reply:
yes, SAN verification on the server side is handled by RBAC. On the client side, SAN verification is handled by tls_context.verification_context. That's why we check it differently.